packages/flutter_tools/lib/src/build_system/targets/windows.dart

```diff
@@ -60,13 +62,13 @@ class UnpackWindows extends Target {
     final String engineSourcePath = environment.artifacts
       .getArtifactPath(
         Artifact.windowsDesktopPath,
-        platform: TargetPlatform.windows_x64,
+        platform: targetPlatform,
         mode: buildMode,
       );
     final String clientSourcePath = environment.artifacts
       .getArtifactPath(
         Artifact.windowsCppClientWrapper,
-        platform: TargetPlatform.windows_x64,
+        platform: targetPlatform,
         mode: buildMode,
       );
     final Directory outputDirectory = environment.fileSystem.directory(
```

Comment:
This ephemeral directory should be specialized by the target architecture. Please see the [Ephemeral build folder section](https://docs.google.com/document/d/1yuexY-EtzeVhU3V6RGBII_kH-znLSkQCspzW70jayZk/edit?resourcekey=0-8pQRSXSHQlwOo7QtCVLvBA#heading=h.sv89rb75mwx1) in the design doc.

Reply:
Specializing the ephemeral build directory path will require updating the [`windows/flutter/CMakeLists.txt`](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/templates/app_shared/windows.tmpl/flutter/CMakeLists.txt) file.

Reply:
A good surprise of this development has been than ephemeral folder is correctly updated when switching architecture. So it's not needed to specialize it per arch, to the opposite of what we expected.

Reply:
You can clone this PR, run both builds arm64 and x64, and you'll notice how files are updated accordingly.

Reply:
I discussed this with co-workers and we're concerned that reusing the ephemeral directory like this could result in difficult bugs in the future. This class of issues can be completely avoided by specializing the ephemeral directory on the target architecture.

I think the best way forward would be to disallow cross-compilation for now (ie: the `--target-platform` value - if provided - must match the host architecture on all release channels). This way we can land this PR mostly unchanged. Once the ephemeral directory is specialized, we can remove this check and allow cross-compilation on the `master` release channel. What are your thoughts?

Reply:
I don't have context on the ephemeral directory issues (but I trust Loic's judgment), but I generally agree with the strategy on difficult PRs like this to cut scope on any ambiguous features that might lead to the whole thing being reverted (which would be bad for the author but also anyone who lands future changes on any of the touched lines of code).

We can then re-land cross-compile in a smaller PR, and discussion on that wouldn't block the rest of this PR that we agree on from landing.

Reply:
If we look more closely at ephemeral folder, we can see it contains:
- `cpp_client_wrapper` source
- `generated_config.cmake`
- copy of flutter engine

Both `cpp_client_wrapper` and cmake generated config do not depend on target platform, they are architecture agnostic.

For flutter engine, we can see in this [test](https://github.com/flutter/flutter/blob/32d42b63def2ea2ab358ff554c959aa02ac4e8ec/packages/flutter_tools/test/general.shard/build_system/targets/windows_test.dart#L58) that it's updated by [`UnpackWindows`](https://github.com/flutter/flutter/blob/32d42b63def2ea2ab358ff554c959aa02ac4e8ec/packages/flutter_tools/lib/src/build_system/targets/windows.dart#L54) function. This [calls](https://github.com/flutter/flutter/blob/32d42b63def2ea2ab358ff554c959aa02ac4e8ec/packages/flutter_tools/lib/src/build_system/targets/windows.dart#L80) `unpackDesktopArtifacts`, with [`_kWindowsArtifacts`](https://github.com/flutter/flutter/blob/32d42b63def2ea2ab358ff554c959aa02ac4e8ec/packages/flutter_tools/lib/src/build_system/targets/windows.dart#L18) as input, and ephemeral directory as output. Finally, the [`unpackDesktopArtifacts`](https://github.com/flutter/flutter/blob/32d42b63def2ea2ab358ff554c959aa02ac4e8ec/packages/flutter_tools/lib/src/build_system/targets/desktop.dart#L16) method will copy inconditionnally artifacts in destination, which is ephemeral folder in our case. From this code analysis, we can understand why it works currently.

---

That said, I'm not a member of the team, and I respect your concern. In this case, we can change the scope of this PR to simply do native compilation (x64 -> x64, arm64 -> arm64).
With that, we could remove the `--target-platform` option, and solve the problem (not yet answered) about which artifacts we should download, by simply getting host ones. There is no more need to check current channel anywhere, as only master will have native dart, thus enabling windows-arm64 native compilation.

Then, later, it could be possible to introduce ephemeral directory specialization, and finally, cross-compilation.

Is that an acceptable plan for you?
Would you like this to be done on another PR?

Reply:
Thank you for the detailed code analysis. Even if the code works today, shared ephemeral directories might result in subtle bugs in the future as we continue to evolve the tool. We'd need to design around this for any future features.

Another reason for specializing the ephemeral directory: after you've `flutter run` or `flutter build` once, you can build a Flutter Windows app using CMake or Visual Studio directly. For example, you can open the `build\windows\x64\my_app.sln` file using Visual Studio. Without a specialized ephemeral directory, a `flutter build --target-platform windows-arm64` would break that x64 Visual Studio solution.

> In this case, we can change the scope of this PR to simply do native compilation (x64 -> x64, arm64 -> arm64).
With that, we could remove the --target-platform option, and solve the problem (not yet answered) about which artifacts we should download, by simply getting host ones. There is no more need to check current channel anywhere, as only master will have native dart, thus enabling windows-arm64 native compilation.
>
> Then, later, it could be possible to introduce ephemeral directory specialization, and finally, cross-compilation.

We should still restrict Windows Arm64 targeting to the master channel. Dart's Windows Arm64 support might be ready for beta/stable channels before Flutter's Windows Arm64 support. We should be able to enable Flutter's Windows Arm64 support separately from Dart.

Otherwise, this sounds perfect 👍 

> Would you like this to be done on another PR?

Personally, I would split the ephemeral directory specialization and cross-compilation to one or more separate PRs, but otherwise reuse this existing PR.

Reply:
I updated PR to only target native compilation. We can continue discussion in normal thread.